### PR TITLE
fix(runtime): cancel linux, darwin, and js builds

### PR DESCRIPTION
**Pull request by m-hilgendorf**
Cancelling a build now makes sure the JS, MacOS, and Linux runtimes clean up after themselves. 

- Linux: The root process of the Linux sandbox is killed, and any children within the corresponding pid namespace created by the guest process are also killed. 
- MacOS: After receiving the stop signal on MacOS, child pids are collected recursively and subsequently reaped.
- JS: Builds are canceled when they yield control back to Rust from JS. This has the downside that CPU bound JS builds will not be canceled until they yield control.
- In a future PR, all JS builds may get their own isolate in their own thread which will support cancellation of CPU bound JS builds.

### CODE REVIEW

File: packages/client/src/build.rs [hunk 1]
```diff
@@ -364,6 +364,38 @@ impl Outcome {
...
+
+	pub fn terminated(&self) -> bool {
+		if let Outcome::Terminated = self {
+			true
+		} else {
+			false
+		}
+	}
+
+	pub fn canceled(&self) -> bool {
```
Comment (m-hilgendorf) [inline packages/client/src/build.rs]: All these methods are added for completeness.

File: packages/runtime/src/darwin.rs [hunk 2]
```diff
@@ -18,8 +19,9 @@ pub async fn build(
 	tg: &dyn tg::Handle,
 	build: &tg::Build,
 	_retry: tg::build::Retry,
+	mut stop: tokio::sync::watch::Receiver<bool>,
 	server_directory_path: &Path,
-) -> Result<tg::Value> {
+) -> Result<Option<tg::Value>> {
```
Comment (m-hilgendorf) [inline packages/runtime/src/darwin.rs:24]: Canceled builds return `Ok(None))`

File: packages/client/src/build.rs (hunk 1 above)
Comment (deciduously) [reply packages/client/src/build.rs]: Could use `matches!()`?

Comment (nitsky) [reply packages/client/src/build.rs]: Agreed, these methods are unnecessary. I think to be idiomatic it should be `is_canceled` etc.

**Pull request merged**